### ISSUE TITLE
ratelimit -ing based on headers. 

### ISSUE THREAD

**Issue by techsin** [trevorsenior/snoocore#115]
Hi I know you can set throttle, but there should also be a way to rate limit and call some function if limit is reached. Thanks. 

Ratelimit is presented in response headers.

**Comment by trevorsenior** [comment]
Going off of this: https://www.reddit.com/r/redditdev/comments/1yxrp7

Headers:

- `X-Ratelimit-Used`: Approximate number of requests used in this period
- `X-Ratelimit-Remaining`: Approximate number of requests left to use
- `X-Ratelimit-Reset`: Approximate number of seconds to end of period

> there should [..] be a way to rate limit [..] if limit is reached

This is a sensible default. It will only be needed if someone sets the throttle to a value less than 1second/request for OAuth clients.

e.g. Lets say someone disabled throttle completly, `throttle: 0`. We will allow them to hit reddit 60 times, reach the reddit rate limit, and then delay the next call for the amount of seconds specified in `X-Ratelimit-Reset`.

However I don't want this library to have strict rules (even if this is a sensible default). Therefore, a new options will be added that will disable this feature & allow for someone to hit reddit as much as they please:

```javascript
honorRatelimitHeaders: true // default - set to false to disable.
```

> there should [..] be a way to [..] call some function if limit is reached. 

A more general solution (using events) could work for this. You would be able to hook into the event and get the information back on each call:

```javascript 
// event is fired on every response from reddit
reddit.on('rate_limit', function(data) {
  // responds with the header data returned from reddit on
  // every response
  console.log(data.ratelimitUsed, 
                     data.ratelimitRemaining, 
                     data.ratelimitReset);
}); 

reddit.on('rate_limit_reached', function(data) {
  // ratelimitRemaining <= 0
});
```

This part will need more thought. Please share any input on this @techsin 

- - -

This will be addressed in `3.0.0` (or in `3.1.0`) as it's near completion, and the library has changed drastically between 2.x.x and 3.x.x!

**Comment by techsin** [comment]
Just to be clear on somethings...

Does snoocore right now checks rateLimitRemaining before making requests?  if no then look Section A?

As a developer i'd want snoocore to by default take care of not breaking api rules. And it's ok if there is a way to over ride.

Having events like that are also a great feature as it'd come in handy in some kind of application flow. But still snoocore should be responsible for dealing throttling if limit is reached.

I've a question do api limits apply to server, app, or user. If to server than that's that. If app then it should be possible to bypass that using multiple api keys, and if user then multiple users.
Maybe snoocore should allow a way for devs to provide a pool of api keys. If it can work.

The whole reason ratelimit feature i'm asking is because if this code is running in browser and if all users are throttled even then limit can easily be surpassed if say 500 users make continuous requests. 

While we are at it, i want to ask you something as you seem to be very familiar with reddit api.
Is it true that certain things have limit of 1000. live saveds. Meaning you can only get last 1000 then reddit won't show it. 
And if yes is there a way to tell reddit api to give 1000 after first 1000. from 1001 to 2000

**Comment by trevorsenior** [comment]
> Does snoocore right now checks rateLimitRemaining before making requests?

You're right. It currently does not look at these headers & sticks to 1 request / second. This makes it impossible to go over the reddit API rules, but does not allow "burst" requests which this feature would allow (and would be really useful!).

> As a developer i'd want snoocore to by default take care of not breaking api rules.

It strictly follows the API rules. You can think of a single instance (`new Snoocore`) as a single users reddit account. Even if you had 500 requests on a single Snoocore instance it will not call the reddit api 500 times at once. It will instead:

- Call an endpoint
- wait 1s
- Call an endpoint
- wait 1s
- ... repeat until all 500 endpoints have been called.

This will be slow if every user shared a single instance so don't do this!

> I've a question do api limits apply to server, app, or user. If to server than that's that. If app then it should be possible to bypass that using multiple api keys, and if user then multiple users.

It is user based.

If you authenticate a Snoocore instance with a user, the rate limits will only be for each user.

Feel free to make several Snoocore instances per user (store them in a hash / etc). I understand that is is a common use case for developers writing web applications, and plan on writing a wrapper around Snoocore that will make this easier.

Here is an example of how you might go about managing multiple Snoocore instances:

```javascript
var userAccount = {};

userAccount['one'] = new Snoocore(/* ... */); // has it's own throttle
userAccount['two'] = new Snoocore(/* ... */); // has it's own throttle
// ...

// some example where a user makes a request, we get their account identifier
// and then we make a call using their Snoocore instance
app.get('/me', function(req, res) {
   var acct = req.cookies['acct'];
   userAccount[acct]('/api/v1/me').get().then(req.send);
});
```

With the above, every *user* has their own Snoocore instance. All instances share the same OAuth key / secrets, however each instance it authenticated with a different user so this is okay!

**please note!**

The above requires that you're authenticated with a user. [Application only OAuth](https://github.com/reddit/reddit/wiki/OAuth2#application-only-oauth) is coming in 3.0.0 which will allow us to do the same thing for unauthenticated users (but it will still require a unique identifier / multiple Snoocore instances)

>While we are at it, i want to ask you something as you seem to be very familiar with reddit api.
Is it true that certain things have limit of 1000. live saveds. Meaning you can only get last 1000 then reddit won't show it. 

Just wrote a test for this:

```javascript
var i = 0;
var total = 0;

reddit('/hot').listing({ limit: 100 }).then(loop);

function loop(slice) {
  if (slice.children.length === 0) {
    console.log('no more!');
    return;
  }

  total += slice.children.length;
  i++
  console.log(i, 'children: ' + slice.children.length, 'total: ' + total);
  return slice.next().then(loop);
}
```

and the results

```
1 'children: 100' 'total: 100'
2 'children: 100' 'total: 200'
3 'children: 100' 'total: 300'
4 'children: 100' 'total: 400'
5 'children: 100' 'total: 500'
6 'children: 100' 'total: 600'
7 'children: 100' 'total: 700'
8 'children: 100' 'total: 800'
9 'children: 100' 'total: 900'
10 'children: 93' 'total: 993'
no more!
```

So yes, it seems like it is limited to the first 1000 entries for a listing.

> And if yes is there a way to tell reddit api to give 1000 after first 1000. from 1001 to 2000

I'm not sure, but can look into it later. This may be a better question for the reddit dev subreddit/IRC

**Comment by techsin** [comment]
Hi, Thanks for responding.

I was worrying that if 300 users are on my site. And snoocore is being used
in browser. Then limit of api calls would be broken. As snoocore is just
throttling per user. 500ms per request etc.
Then from my api key there would be 300 requests per second even though
every js instance waited 300ms per request in their respective users'
browsers.

However, as you say, that it's not based on my dev account or apiKey/app
but based on individual user. Right? So even if there are 1000 active users
and all make request simultaneously then it won't break the api rule. Since
per user it's still 1 req/s even though per app and or my dev acct its 1000
req/s.


about limit i'm researching ill let you know if i find definite answer.
Also i think count parameter might have to do something with it.

**Comment by trevorsenior** [comment]
Ah, so this is a client side app? 100% browser based?

Curious to find your results of the 1000 limit / if you get an answer please post back :smile:

**Comment by techsin** [comment]
Sure..

So that's my question.. does it, as of now, checks if ratelimit is breached
or not before making a request. Or it just throttles individual user ? and
can that result in breaking api rules. (it shouldn't right because as you
said it is based per user or auth Token)

it should go like this

1. check current rate limit
2. if not breached or not near it make request

**Comment by trevorsenior** [comment]
Ah sorry, it's hard to diagnose the problem over GitHub at times.

> So that's my question.. does it, as of now, checks if ratelimit is breached
or not before making a request.

No it does not check before making a request. It sits in a queue of calls that are dispatched every second. This falls within the reddit api rules for single user use (even when not authenticated with a user).

> Or it just throttles individual user ?

Each new instance of Snoocore has an internal timer/throttle. If you use one instance per user, then yes. It just throttles for the individual user.

> and can that result in breaking api rules.

**Yes it can break api rules depending on use case**

1. *Unauthenticated Users* -- If you have many Snoocore instances, all **unauthenticated** making thousands of calls to the reddit API at once, reddit views these as coming from the same source and this is against the rules.

2. *Authenticated Users* -- However, if you **authenticate with a user** for *every* Snoocore instance and make those same calls, this is *allowed*!
- - -

In Snoocore 3.x.x the first case above will be solved with Application only OAuth.

In version 2.x.x there is no way to identify unauthenticated users uniquely. 

- - -

All in all, I recommend requiring that *all* users log in with their reddit account before they user your app. 

If that is not an option, I'm working on finishing 3.x.x as soon as I can. There are some technical road blocks with the completion of it. Namely, I'm waiting on a fix from reddit for (https://github.com/trevorsenior/snoocore/issues/94#issuecomment-83711917).

**Comment by techsin** [comment]
No getting auth is no problem all users are authd. Thanks i love your lib.

**Comment by trevorsenior** [comment]
Ah great. As long as they are authenticated you're in the clear :wink: 

I'm closing this issue, but I've opened another issue, #116, that will add in the rate limit headers. Thanks for bringing it up.
